source/extensions/filters/network/redis_proxy/command_splitter_impl.cc

```diff
@@ -54,20 +70,61 @@ void SingleServerRequest::onFailure() {
   callbacks_.onResponse(Utility::makeError(Response::get().UpstreamFailure));
 }
 
+bool SingleServerRequest::onMovedRedirection(Common::Redis::RespValue& value) {
+  std::vector<absl::string_view> err;
+  if (redirectionArgsInvalid(incoming_request_, value, err)) {
+    return false;
+  }
+
+  const std::string host_address = std::string(err[2]); // ip:port
+  handle_ = conn_pool_.makeRequestToHost(host_address, *incoming_request_, *this);
+  return (handle_ != nullptr);
+}
+
+bool SingleServerRequest::onAskRedirection(Common::Redis::RespValue& value) {
+  std::vector<absl::string_view> err;
+  if (redirectionArgsInvalid(incoming_request_, value, err)) {
+    return false;
+  }
+
+  Common::Redis::RespValue request_with_asking;
+
+  if (incoming_request_->asArray().size() > 0 &&
+      incoming_request_->type() == Common::Redis::RespType::Array) {
+
+    Common::Redis::RespValue asking_cmd;
+    asking_cmd.type(Common::Redis::RespType::BulkString);
+    asking_cmd.asString() = "asking";
+    request_with_asking.type(Common::Redis::RespType::Array);
+    request_with_asking.asArray().push_back(asking_cmd);
+    request_with_asking.asArray().insert(request_with_asking.asArray().end(),
+                                         incoming_request_->asArray().begin(),
+                                         incoming_request_->asArray().end());
+  } else {
+    return false;
+  }
+
+  const std::string host_address = std::string(err[2]); // ip:port
+  handle_ = conn_pool_.makeRequestToHost(host_address, request_with_asking, *this);
+  return (handle_ != nullptr);
+}
+
 void SingleServerRequest::cancel() {
   handle_->cancel();
   handle_ = nullptr;
 }
 
 SplitRequestPtr SimpleRequest::create(ConnPool::Instance& conn_pool,
-                                      const Common::Redis::RespValue& incoming_request,
+                                      Common::Redis::RespValuePtr&& incoming_request,
                                       SplitCallbacks& callbacks, CommandStats& command_stats,
                                       TimeSource& time_source, bool latency_in_micros) {
-  std::unique_ptr<SimpleRequest> request_ptr{
-      new SimpleRequest(callbacks, command_stats, time_source, latency_in_micros)};
+  const Common::Redis::RespValue* incoming_request_ptr = incoming_request.get();
```

Comment:
This pattern should be avoided if possible. Why do you need to move the incoming request vs. passing as a const ref as it is now? Mainly trying to understand so we can determine if there is a better way to do this.

Reply:
If the incoming request is passed in as a const ref as it is now, then the actual RespValue gets freed after the stack unwinds after the buffer encoding of the value. This code saves off the pointer. I'll be making some changes, so that incoming_request_ is not set in the constructor. It'll make everything a little less awkward, and clearer. Stay tuned.

Reply:
the changes are up -- incoming_request is saved away in a request unique pointer, incoming_request_, only at the end of the static create() calls